Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducer.swift

```diff
@@ -0,0 +1,1064 @@
+                    return element
+
+                case .returnElementAndCallProduceMore(let element):
+                    let delegate = self.delegate
+                    self.lock.unlock()
+
+                    delegate?.produceMore()
+
+                    return element
+
+                case .returnElementAndCallDidTerminate(let element):
+                    let delegate = self.delegate
+                    self.delegate = nil
+                    self.lock.unlock()
+
+                    delegate?.didTerminate()
+                    return element
+
+                case .returnNil:
+                    self.lock.unlock()
+                    return nil
+
+                case .suspendTask:
+                    // It is safe to hold the lock across this method
+                    // since the closure is guaranteed to be run straight away
+                    return await withCheckedContinuation { continuation in
+                        let action = self.stateMachine.next(for: continuation)
+
+                        switch action {
+                        case .callProduceMore:
+                            let delegate = delegate
+                            self.lock.unlock()
+
+                            delegate?.produceMore()
+
+                        case .none:
+                            self.lock.unlock()
+                        }
+                    }
+                }
+            } onCancel: {
+                let delegate: Delegate? = self.lock.withLock {
+                    let action = self.stateMachine.cancelled()
+
+                    switch action {
+                    case .callDidTerminate:
+                        let delegate = self.delegate
+                        self.delegate = nil
+
+                        return delegate
+
+                    case .resumeContinuationWithNilAndCallDidTerminate(let continuation):
+                        continuation.resume(returning: nil)
+                        let delegate = self.delegate
+                        self.delegate = nil
+
+                        return delegate
+
+                    case .none:
+                        return nil
+                    }
+                }
+
+                delegate?.didTerminate()
+            }
+        }
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncSequenceProducer {
+    @usableFromInline
+    /* private */ internal struct StateMachine {
+        @usableFromInline
+        /* private */ internal enum State {
+            /// The initial state before either a call to `yield()` or a call to `next()` happened
+            case initial(
+                backPressureStrategy: Strategy,
+                iteratorInitialized: Bool
+            )
+
+            /// The state once either any element was yielded or `next()` was called.
+            case streaming(
+                backPressureStrategy: Strategy,
+                buffer: Deque<Element>,
```

Comment:
I discussed this with @Lukasa yesterday and he was fine in bringing in this dependency. The back story is that `Deque` is more performant than `CiruclarBuffer` since it implements a few customization points.